nupic/research/frameworks/pytorch/datasets/permuted_mnist.py

```diff
@@ -135,6 +173,46 @@ def init_contexts(self, seed):
             self.contexts[i, :] = self.contexts[i, torch.randperm(self.dim_context,
                                                                   generator=g)]
 
+    def init_centroids(self):
+        """
+        Code to compute the mean image from each permutation. Note that you only need
+        to compute the mean image for the base dataset. After that you can just apply
+        each permutation to the mean vector.
+        """
+        self.dim_context = 784
+        self.centroids = torch.zeros((self.num_tasks, 28, 28))
+        for index in range(len(self.data)):
+            img, _ = super().__getitem__(index)
+            self.centroids[0] += img.squeeze(0)
+
+        # This first row has the pixelwise sum for MNIST, divide to get mean
+        self.centroids[0] /= len(self.data)
+
+        # Now just apply permutations to the mean vector, one for each remaining task
+        for task in range(1, self.num_tasks):
+            self.centroids[task] = permute(self.centroids[0].unsqueeze(0),
+                                           self.permutations[task])
+
+        # 28 x 28 -> 784
+        self.contexts = self.centroids.flatten(start_dim=1)
+
+    def init_one_hot_context(self):
+        self.dim_context = self.num_tasks
+        self.contexts = torch.eye(self.num_tasks, self.num_tasks)
+
+
+class SelfContextPermutedMNIST(PermutedMNIST):
+
+    def __init__(self, num_tasks, seed, train, root=".", target_transform=None,
+                 download=False):
```

Comment:
Good point that it shouldn't be in a config. I think you can just create a mixin with a few lines of code (!) that does the trick, and would be general for any dataset compatible with Vernon, not just permutedMNIST. It would look like

```
class SelfContext:

    def setup_experiment(self, config):
        super().setup_experiment(config)

        self.train_context_fn = self_context
        self.infer_context_fn = self_context


def self_context(data):
    return data
```

But I don't see why you would need to change centroid_context. The above solution would be compatible with that.